### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [390]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

## Player Count

* Display the total number of players


In [391]:
#A check to see if reference data is opened properly
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [392]:
#Checks data information
purchase_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 780 entries, 0 to 779
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Purchase ID  780 non-null    int64  
 1   SN           780 non-null    object 
 2   Age          780 non-null    int64  
 3   Gender       780 non-null    object 
 4   Item ID      780 non-null    int64  
 5   Item Name    780 non-null    object 
 6   Price        780 non-null    float64
dtypes: float64(1), int64(3), object(3)
memory usage: 42.8+ KB


In [393]:
#This checks to see if there are duplicates players in the SN column and if so, how many
purchase_data.SN.duplicated().sum()

#Alternatively, this code can be used to view all the players that are duplicated.
#it shows the information only for the first occurence for each player 
#purchase_data_duplicates = purchase_data.loc[purchase_data.SN.duplicated(keep='first'),:]
#purchase_data_duplicates

#Since duplicate players exist, this code removes the duplicates and counts the unique players
#this is saved as a new variable so as not to mess the original dataframe
purchase_data_no_dups=purchase_data.SN.drop_duplicates()

purchase_data_no_dups

total_players=len(purchase_data_no_dups)
total_players

Total_Players=pd.DataFrame({"Total Players":[total_players]})
Total_Players

Total Players
0            576

## Purchasing Analysis (Total)

* Run basic calculations to obtain number of unique items, average price, etc.


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [594]:
#To find the number of unique game titles sold
#This loops throught the dataframe and puts the unique names into a list (items_bought), then increases the number of
#unique items by 1 and prints the final number
number_of_unique_items = 0
items_bought = []
for i in range(0,len(purchase_data['Item Name'])):
    if purchase_data['Item Name'][i] not in items_bought:
        items_bought.append(purchase_data['Item Name'][i])
        number_of_unique_items += 1
number_of_unique_items

#This code find the average and formats to be in currency notation
avg_price=purchase_data.Price.mean()
avg_price="${:,.2f}".format(avg_price)
avg_price

#takes the lenth of the data entries to find the amount of purchases
total_purchases = len(purchase_data['Price'])
total_purchases

#sums the Price column to find the total revenue of sales
total_revenue = sum(purchase_data['Price'])
total_revenue="${:,.2f}".format(total_revenue)
total_revenue

#creates new dataframe from calculated values
Purchasing_Analysis_Total=pd.DataFrame({'Number of Unique Items':[number_of_unique_items],
                                       'Average Price':[avg_price],
                                       'Number of Purchases':[total_purchases],
                                       'Total Revenue':[total_revenue]})
Purchasing_Analysis_Total

Number of Unique Items Average Price  Number of Purchases Total Revenue
0                     179         $3.05                  780     $2,379.77

## Gender Demographics

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed




In [399]:
purchase_data.head()

Purchase ID             SN  Age Gender  Item ID  \
0            0        Lisim78   20   Male      108   
1            1    Lisovynya38   40   Male      143   
2            2     Ithergue48   24   Male       92   
3            3  Chamassasya86   24   Male      100   
4            4      Iskosia90   23   Male      131   

                                   Item Name  Price  
0  Extraction, Quickblade Of Trembling Hands   3.53  
1                          Frenzied Scimitar   1.56  
2                               Final Critic   4.88  
3                                Blindscythe   3.27  
4                                       Fury   1.44

In [593]:
#pulls data for each player and their gender
gender_df=purchase_data.loc[:,['SN','Gender']]
#since there are duplicates, this drops the duplicated player data
gender_no_dups=gender_df.drop_duplicates()
#this provides the counts for each gender
gender_counts=gender_no_dups.Gender.value_counts()
gender_counts

#counts each individual sum for female, male, and other players
male=gender_counts.Male.sum()
male
female=gender_counts.Female.sum()
female
other=gender_counts['Other / Non-Disclosed'].sum()
other

#pulls data from the gender and counts the percentage
percentage_of_players=(gender_counts/len(gender_no_dups))*100
percentage_of_players
#counts each individual percentage for female, male, and other players
male_percent="{:.2f}%".format(male/len(gender_no_dups)*100)
male_percent
female_percent="{:.2f}%".format(female/len(gender_no_dups)*100)
female_percent
other_percent="{:.2f}%".format(other/len(gender_no_dups)*100)
other_percent

#creates new dataframe from calcualted values
gender_demographics=pd.DataFrame({'Gender':['Male','Female','Other / Non-Disclosed'],
                                  'Total Count':[male,female,other],
                                  'Percentage of Players':[male_percent,female_percent,other_percent]})

gender_demographics=gender_demographics.set_index("Gender")
gender_demographics#pulls data from the gender and counts the percentage
percentage_of_players=(gender_counts/len(gender_no_dups))*100
percentage_of_players
#counts each individual percentage for female, male, and other players
male_percent="{:.2f}%".format(male/len(gender_no_dups)*100)
male_percent
female_percent="{:.2f}%".format(female/len(gender_no_dups)*100)
female_percent
other_percent="{:.2f}%".format(other/len(gender_no_dups)*100)
other_percent

#creates new dataframe from calcualted values
gender_demographics=pd.DataFrame({'Gender':['Male','Female','Other / Non-Disclosed'],
                                  'Total Count':[male,female,other],
                                  'Percentage of Players':[male_percent,female_percent,other_percent]})

gender_demographics=gender_demographics.set_index("Gender")
gender_demographics

Total Count Percentage of Players
Gender                                                  
Male                           484                84.03%
Female                          81                14.06%
Other / Non-Disclosed           11                 1.91%


## Purchasing Analysis (Gender)

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender




* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [403]:
#extracts data for gender and price
gender_purchases_df=purchase_data.loc[:,['Gender','Price']]

#check the counts for all genders
gender_purchases=gender_purchases_df.Gender.value_counts()
gender_purchases

#counts each individual purchase count for female, male, and other players
male_purchases=gender_purchases.Male.sum()
male_purchases
female_purchases=gender_purchases.Female.sum()
female_purchases
other_purchases=gender_purchases['Other / Non-Disclosed'].sum()
other_purchases

#finds the price total spent for male, female, and other players

male_total_price = gender_purchases_df.loc[gender_purchases_df['Gender']=='Male', ['Price']].sum()
female_total_price
female_total_price = gender_purchases_df.loc[gender_purchases_df['Gender']=='Female', ['Price']].sum()
female_total_price
other_total_price = gender_purchases_df.loc[gender_purchases_df['Gender']=='Other / Non-Disclosed', ['Price']].sum()
other_total_price

#finds the average total price per male, female, and other

male_avg_price= male_total_price/male_purchases
female_avg_price= female_total_price/female_purchases
other_avg_price= other_total_price/other_purchases

#finds average total purchase per male, female, and other

price_per_male = male_total_price/male
price_per_male
price_per_female = female_total_price/female
price_per_female
price_per_other = other_total_price/other
price_per_other

#create dataframe for purchasing analysis
purchasing_analysis = pd.DataFrame({'Gender':['Male','Female','Other / Non-Disclosed'],
                                    'Purchase Count':[male_purchases,female_purchases,other_purchases],
                                    'Average Purchase Price':["${:.2f}".format(male_avg_price[0]),"${:.2f}".format(female_avg_price[0]),"${:.2f}".format(other_avg_price[0])],
                                    'Total Purchase Value':["${:.2f}".format(male_total_price[0]),"${:.2f}".format(female_total_price[0]),"${:.2f}".format(other_total_price[0])],
                                    'Average Total Purchase per Person':["${:.2f}".format(price_per_male[0]),"${:.2f}".format(price_per_female[0]),"${:.2f}".format(price_per_other[0])]})
                              
                               
purchasing_analysis

Gender  Purchase Count Average Purchase Price  \
0                   Male             652                  $3.02   
1                 Female             113                  $3.20   
2  Other / Non-Disclosed              15                  $3.35   

  Total Purchase Value Average Total Purchase per Person  
0             $1967.64                             $4.07  
1              $361.94                             $4.47  
2               $50.19                             $4.56

## Age Demographics

* Establish bins for ages


* Categorize the existing players using the age bins. Hint: use pd.cut()


* Calculate the numbers and percentages by age group


* Create a summary data frame to hold the results


* Optional: round the percentage column to two decimal points


* Display Age Demographics Table


In [450]:
#create bins to place ages and their labels
bins_age=[0,9,14,19,24,29,34,39,44]
bin_labels=['<10','10-14','15-19','20-24','25-29','30-34','35-39','40+']

#places new data into original dataframe

new_purchase_data['Age Groups']=pd.cut(new_purchase_data['Age'], bins=bins_age,labels=bin_labels)
age_purchase_data=new_purchase_data

#pulls data for each player and their gender
age_purchase_count_df=new_purchase_data.loc[:,['Purchase ID','SN','Age','Age Groups','Price']]

#since there are duplicates, this drops the duplicated player data
age_purchase_no_dups=age_purchase_count_df.drop_duplicates('SN')

age_purchase_data=age_purchase_no_dups.groupby("Age Groups")
age_purchase_data.count()

#puts data into dataframe
purchases_by_age=pd.DataFrame(age_purchase_data.count())
purchases_by_age


purchases_by_age["Purchase ID"]=round(purchases_by_age["Purchase ID"]/total_players * 100,2)

purchases_by_age["Purchase ID"]=purchases_by_age["Purchase ID"].map("{:,.2f}%".format)
purchasing_age_analysis=purchases_by_age[["Purchase ID","SN"]]
purchasing_age_analysis=purchasing_age_analysis.rename(columns={"Purchase ID":"Percentage of Players","SN":"Total Count"})
purchasing_age_analysis

Percentage of Players  Total Count
Age Groups                                   
<10                        2.95%           17
10-14                      3.82%           22
15-19                     18.58%          107
20-24                     44.79%          258
25-29                     13.37%           77
30-34                      9.03%           52
35-39                      5.38%           31
40+                        1.91%           11

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age


* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below


* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame

In [349]:
#checks the number of people for each age group
avg_tot_per_person

Age Groups
<10      3.35
10-14    2.96
15-19    3.04
20-24    3.05
25-29    2.90
30-34    2.93
35-39    3.60
40+      3.04
dtype: float64

In [509]:
#sort data by age groups
purchase_analysis=purchase_data
purchase_analysis_data=purchase_analysis.groupby("Age Groups")
purchase_by_age_analysis=pd.DataFrame(purchase_analysis_data["Purchase ID"].count())
purchase_by_age_analysis

#calculate average purchase price
ave_age_purchase=purchase_analysis_data["Price"].mean()

#calclates total purhcase value
tot_purchase_age=purchase_analysis_data["Price"].sum()

#calculates average total purchase per person
avg_tot_per_person=tot_purchase_age/purchasing_age_analysis['Total Count']

#put all calculated points into dataframe and format numbers to currency
purchase_by_age_analysis['Average Purchase Price'] = ave_age_purchase
purchase_by_age_analysis['Average Purchase Price'] = purchase_by_age_analysis['Average Purchase Price'].map("${:,.2f}".format)

purchase_by_age_analysis['Total Purchase Value'] = tot_purchase_age
purchase_by_age_analysis['Total Purchase Value'] = purchase_by_age_analysis['Total Purchase Value'].map("${:,.2f}".format)
purchase_by_age_analysis['Avg Total Purchase per Person']=avg_tot_per_person
purchase_by_age_analysis['Avg Total Purchase per Person'] = purchase_by_age_analysis['Avg Total Purchase per Person'].map("${:,.2f}".format)
purchase_by_age_analysis=purchase_by_age_analysis.rename(columns={"Purchase ID":"Purchase Counts"})
purchase_by_age_analysis

Purchase Counts Average Purchase Price Total Purchase Value  \
Age Groups                                                                
<10                      23                  $3.35               $77.13   
10-14                    28                  $2.96               $82.78   
15-19                   136                  $3.04              $412.89   
20-24                   365                  $3.05            $1,114.06   
25-29                   101                  $2.90              $293.00   
30-34                    73                  $2.93              $214.00   
35-39                    41                  $3.60              $147.67   
40+                      13                  $2.94               $38.24   

           Avg Total Purchase per Person  
Age Groups                                
<10                                $4.54  
10-14                              $3.76  
15-19                              $3.86  
20-24                              $4.32  
25-29                              $3.81  
30-34                              $4.12  
35-39                              $4.76  
40+                                $3.48

## Top Spenders

* Run basic calculations to obtain the results in the table below


* Create a summary data frame to hold the results


* Sort the total purchase value column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [582]:
spender_purchase_data_df=pd.DataFrame(purchase_data)

#organizes data by buyer and put it in a dataframe
buyer_data=spender_purchase_data_df.groupby("SN")
buyer_data.count()
buyer_analysis_df=pd.DataFrame(buyer_data["Purchase ID"].count())
buyer_analysis_df


#average purchases by buyer
avg_buyer_purchase=buyer_data["Price"].mean()
avg_buyer_purchase


#total price of overall purchases
purchase_by_buyer=buyer_data["Price"].sum()
purchase_by_buyer


#put information into dataframe
buyer_analysis_df["Average Purchase Price"]=avg_buyer_purchase
buyer_analysis_df["Total Purchase Value"]=purchase_by_buyer
buyer_analysis_df=buyer_analysis_df.rename(columns={"Purchase ID":"Purchase Counts"})
buyer_analysis_df

#map("${:,.2f}".format)
#sort data by top spenders
top_spender_summary=buyer_analysis_df.sort_values(by="Total Purchase Value",ascending=False)
top_5_spenders=top_spender_summary
top_5_spenders["Total Purchase Value"]=purchase_by_buyer.map("${:,.2f}".format)
top_5_spenders["Average Purchase Price"]=avg_buyer_purchase.map("${:,.2f}".format)
top_5_spenders.head(5)

Purchase Counts Average Purchase Price Total Purchase Value
SN                                                                      
Lisosia93                  5                  $3.79               $18.96
Idastidru52                4                  $3.86               $15.45
Chamjask73                 3                  $4.61               $13.83
Iral74                     4                  $3.41               $13.62
Iskadarya95                3                  $4.37               $13.10

## Most Popular Items

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, average item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [586]:
#sort by items
top_items_df=purchase_data.groupby(["Item ID","Item Name"])

#create new dataframe from sorted items
item_analysis_df=pd.DataFrame(top_items_df["Purchase ID"].count())

#calculate item price
items_price=top_items_df["Price"].mean()

#calulate total purchase value
item_total_purchase=top_items_df["Price"].sum()

#put data into dataframe
item_analysis_df["Item Price"]=items_price.map("${:,.2f}".format)
item_analysis_df["Total Purchase Value"]=item_total_purchase.map("${:,.2f}".format)
item_analysis_df=item_analysis_df.rename(columns={"Purchase ID":"Purchase Counts"})

#sort by items with most sales
top_5_item_analysis_df=item_analysis_df.sort_values("Purchase Counts",ascending=False)
top_5_item_analysis_df.head(5)

Purchase Counts  \
Item ID Item Name                                                       
92      Final Critic                                               13   
178     Oathbreaker, Last Hope of the Breaking Storm               12   
145     Fiery Glass Crusader                                        9   
132     Persuasion                                                  9   
108     Extraction, Quickblade Of Trembling Hands                   9   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
145     Fiery Glass Crusader                              $4.58   
132     Persuasion                                        $3.22   
108     Extraction, Quickblade Of Trembling Hands         $3.53   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
145     Fiery Glass Crusader                                       $41.22  
132     Persuasion                                                 $28.99  
108     Extraction, Quickblade Of Trembling Hands                  $31.77

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [591]:
#use the dataframe made above to sort for sum of price
item_analysis_df["Total Purchase Value"]=top_items_df["Price"].sum()

#sort total purchase prices and format prices
top_profitable_items=item_analysis_df.sort_values("Total Purchase Value", ascending=False)
top_profitable_items["Total Purchase Value"]=item_total_purchase.map("${:,.2f}".format)
top_profitable_items.head(5)

Purchase Counts  \
Item ID Item Name                                                       
92      Final Critic                                               13   
178     Oathbreaker, Last Hope of the Breaking Storm               12   
82      Nirvana                                                     9   
145     Fiery Glass Crusader                                        9   
103     Singed Scalpel                                              8   

                                                     Item Price  \
Item ID Item Name                                                 
92      Final Critic                                      $4.61   
178     Oathbreaker, Last Hope of the Breaking Storm      $4.23   
82      Nirvana                                           $4.90   
145     Fiery Glass Crusader                              $4.58   
103     Singed Scalpel                                    $4.35   

                                                     Total Purchase Value  
Item ID Item Name                                                          
92      Final Critic                                               $59.99  
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76  
82      Nirvana                                                    $44.10  
145     Fiery Glass Crusader                                       $41.22  
103     Singed Scalpel                                             $34.80

# HeroesOfPymoli Analysis


1)	There was a total of 780 purchases, with some players buying more than one game. Average price of games purchased was $3.05 , total revenue was $2,379.77, and 179 unique titles were sold
2)	Of the 576 players, roughly 84% of the players were male. The remaining 15% is split between females (14.06%) and other/non-disclosed individuals (1.91%)
3)	Due to the majority of the players being males, it makes sense that this is the group that purchases (652 items) with a total purchase revenue of close to $2000. On the otherhand, females purchased 113 items with a total revenue of close to $400..
4)	For the distribution of age, 44.79% of players are between ages 25-29 followed by age group 20-24 (18.58%). Age group 25-30 also had the most sales as well total purchase value.
5)	The top spenders for the games spend an average of $3.79 on the game, while the average price of each item is $3.05. Due to popularity or usefulness of items, it likely that players are willing to spend more than the average cost of an item to improve play quality.
6)	The most popular item bought was the Final Critic with a total purchase value of $59.99. Oathbreaker, Last Hope of the Breaking Storm was just as popular by being behind by one purchase count with $50.76 in sales. These two items were also the most profitable out of all the items sold. 183 different items were sold, yet these two sold the most and made most revenue.
